btree/btree.c

```diff
@@ -3570,6 +3575,67 @@ int64_t btree_open_tick(struct m0_sm_op *smop)
 	}
 }
 
+/**
+ * btree_close_tick function is used to traverse through different states to
+ * facilitate the working of m0_btree_open().
+ *
+ * @param smop     represents the state machine operation
+ * @return int64_t returns the next state to be executed.
+ */
+int64_t btree_close_tick(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop     = M0_AMB(bop, smop, bo_op);
+	struct m0_btree       *tree    = bop->bo_arbor;
+	struct nd             *nd_head = ndlist_tlist_head(&tree->t_desc->
+							   t_active_nds);
+	struct nd             *nd_curr;
+
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+	if (tree->t_desc->t_ref > 1) {
+		tree_put(tree->t_desc);
+	} else if (tree->t_desc->t_ref == 1) {
+		nd_curr = nd_head;
+		if(nd_curr->start_time == 0)
+			nd_curr->start_time = m0_time_now();
+		do {
+			node_put(nd_curr);
+			if (nd_curr->n_ref > 0) {
+				if (nd_curr == tree->t_desc->t_root &&
+				    nd_curr->n_ref == 1) {
+					nd_curr->start_time = 0;
+					node_put(nd_curr);
+				}
+				else {
+					/** This code is meant for debugging.
+					 *  In future, this case needs to be
+					 *  handled in a better way.
+					 */
+					if (m0_time_seconds(m0_time_now() -
+					    nd_curr->start_time) > 5) {
+						nd_curr->start_time = 0;
+						return M0_ERR(-ETIMEDOUT);
+					}
+					return P_INIT;
+				}
+			}
```

Comment:
Why are you putting the nodes. Some other thread might be using them !!!

Reply:
We are putting nodes related to current tree only.
When the tree has t_ref=1 which means it will be released, we need to clear the node descriptors associated with it as well. node_get() will increase the n_ref count. node_put() will decrease it.
 Hence, if after node_put() n_ref is not zero, then we check for the error case.
If n_ref is zero which means the node descriptor is free to be cleared.

Reply:
What if the tree_close is called and all the nodes are not yet released because the thread in which they are running is a bit slow or is getting suspended by linux. In this case the nodes will eventually be 'put' but the assumption of the nodes and the code using the nodes about the existence of tree and the nodes will be violated if you were to put the nodes without their knowledge. You need to wait till the nodes are all put and also till the t_ref goes to 0. You cannot hurry up the other threads as they are not in your control.

Reply:
Yes, on discussing with Nikhil I got the issue.
I have modified the code as per the requirements.